plugins/src/main/java/tech/pegasys/pantheon/plugins/PantheonContext.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.plugins;
+
+import java.util.Optional;
+
+public interface PantheonContext {
+
+  <T> Optional<T> getService(Class<T> serviceType);
```

Comment:
Curious - why go with this generic pattern rather than just explicitly exposing some services like `registerCLIOptions`, `getPantheonEvents`, etc?  Is the idea that this makes it easier for us to remove services in the future?

Reply:
Also wondering if we shouldn't just return the service class directly rather than wrapping in it in an `Optional`?  And in the case where a plugin asks for a service that isn't registered, we could just throw an exception and disable that plugin.  Seems safer to disable plugins that might be missing required services.

Reply:
This is so we can expose more services later and not have to mutate the API.  In a future revision the plugins themselves could expose services.

The Optional is there in case plugins want to do graceful degraded service.  A plugin may want a service but not require it.  For example a plugin that interacts with consensus engines would only expect one to be operating but would go down the list of possible engines that expose services to the plugin architecture.